File: snake3d.py
# snake3d.py
import os
import random
import time
import sys
import threading
from collections import deque

class Snake3D:
    def __init__(self, size=8):
        self.size = size
        self.center = size // 2
        self.snake = deque()
        self.snake.append((self.center, self.center, self.center))
        self.direction = (0, 0, 1)  # x, y, z
        self.next_direction = self.direction
        self.food = self._spawn_food()
        self.score = 0
        self.high_score = 0
        self.speed = 0.3
        self.game_over = False
        self.paused = False
        self.running = True

    def _spawn_food(self):
        while True:
            pos = (random.randint(0, self.size-1),
                   random.randint(0, self.size-1),
                   random.randint(0, self.size-1))
            if pos not in self.snake:
                return pos

    def _project_3d_to_2d(self, x, y, z):
        """Isometric projection: (x,y,z) -> (screen_x, screen_y)"""
        screen_x = (x - z) * 2 + self.size * 2
        screen_y = (x + z) + y * 2 + 2
        return screen_x, screen_y

    def _render(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print(f"🐍 3D SNAKE   Score: {self.score}   High: {self.high_score}   Speed: {self.speed:.2f}s")
        print("Controls: W/A/S/D/Q/E | Space=pause | R=restart | Esc=quit")
        # Build grid view (isometric)
        # We'll render layer by layer from back to front
        layers = {}
        # Add snake segments
        for idx, seg in enumerate(self.snake):
            layers[seg] = 'O' if idx == 0 else 'o'
        # Add food
        layers[self.food] = '★'
        # Render
        for y in range(self.size-1, -1, -1):
            for z in range(self.size):
                line = []
                for x in range(self.size):
                    pos = (x, y, z)
                    if pos in layers:
                        line.append(layers[pos])
                    else:
                        line.append('.')
                print(' '.join(line))
        # Show 3D projection
        print("\n3D Isometric View:")
        screen = [[' ' for _ in range(self.size*4 + 4)] for _ in range(self.size*4 + 4)]
        for y in range(self.size):
            for z in range(self.size):
                for x in range(self.size):
                    pos = (x, y, z)
                    if pos in layers:
                        sx, sy = self._project_3d_to_2d(x, y, z)
                        if 0 <= sx < len(screen[0]) and 0 <= sy < len(screen):
                            screen[sy][sx] = layers[pos]
        for row in screen:
            print(''.join(row))
        if self.paused:
            print("\n⏸️  PAUSED")
        if self.game_over:
            print("\n💀 GAME OVER! Press R to restart.")

    def _update(self):
        if self.paused or self.game_over:
            return
        self.direction = self.next_direction
        head = self.snake[0]
        new_head = (head[0] + self.direction[0],
                    head[1] + self.direction[1],
                    head[2] + self.direction[2])
        # Wrap around world (optional)
        new_head = (new_head[0] % self.size,
                    new_head[1] % self.size,
                    new_head[2] % self.size)
        # Check collision with self
        if new_head in self.snake:
            self.game_over = True
            if self.score > self.high_score:
                self.high_score = self.score
            return
        self.snake.appendleft(new_head)
        if new_head == self.food:
            self.score += 1
            self.food = self._spawn_food()
            self.speed = max(0.08, self.speed - 0.01)
        else:
            self.snake.pop()

    def run(self):
        # Input thread
        def input_loop():
            while self.running:
                try:
                    ch = sys.stdin.read(1)
                    if ch == ' ':
                        self.paused = not self.paused
                    elif ch == 'r' or ch == 'R':
                        self.__init__(self.size)
                    elif ch == '\x1b' and not self.paused and not self.game_over:
                        # Arrow keys: read next two chars
                        c = sys.stdin.read(2)
                        if c == '[A': self.next_direction = (0, 0, 1)
                        elif c == '[B': self.next_direction = (0, 0, -1)
                        elif c == '[D': self.next_direction = (-1, 0, 0)
                        elif c == '[C': self.next_direction = (1, 0, 0)
                    elif ch == 'w' or ch == 'W':
                        self.next_direction = (0, 0, 1)
                    elif ch == 's' or ch == 'S':
                        self.next_direction = (0, 0, -1)
                    elif ch == 'a' or ch == 'A':
                        self.next_direction = (-1, 0, 0)
                    elif ch == 'd' or ch == 'D':
                        self.next_direction = (1, 0, 0)
                    elif ch == 'q' or ch == 'Q':
                        self.next_direction = (0, 1, 0)
                    elif ch == 'e' or ch == 'E':
                        self.next_direction = (0, -1, 0)
                    elif ch == '\x1b':  # Esc key
                        self.running = False
                except:
                    pass
        threading.Thread(target=input_loop, daemon=True).start()
        while self.running and not self.game_over:
            self._update()
            self._render()
            time.sleep(self.speed)
        if self.game_over:
            self._render()
            while self.running:
                ch = sys.stdin.read(1)
                if ch == 'r' or ch == 'R':
                    self.__init__(self.size)
                    self.run()
                    return
                elif ch == '\x1b':
                    break

if __name__ == "__main__":
    game = Snake3D()
    game.run()

File: Snake3D.cs
// Snake3D.cs
using System;
using System.Collections.Generic;
using System.Threading;

class Snake3D
{
    const int SIZE = 8;
    List<(int x, int y, int z)> snake = new List<(int, int, int)>();
    (int x, int y, int z) dir = (0, 0, 1);
    (int x, int y, int z) nextDir = (0, 0, 1);
    (int x, int y, int z) food;
    int score = 0, high = 0;
    int speed = 300; // ms
    bool gameOver = false, paused = false, running = true;
    Random rand = new Random();

    public Snake3D()
    {
        int c = SIZE / 2;
        snake.Add((c, c, c));
        SpawnFood();
        // Input thread
        Thread inputThread = new Thread(InputLoop);
        inputThread.IsBackground = true;
        inputThread.Start();
        Run();
    }

    void SpawnFood()
    {
        while (true)
        {
            var pos = (rand.Next(SIZE), rand.Next(SIZE), rand.Next(SIZE));
            if (!snake.Contains(pos))
            {
                food = pos;
                break;
            }
        }
    }

    bool Contains((int x, int y, int z) pos)
    {
        return snake.Contains(pos);
    }

    void Update()
    {
        if (paused || gameOver) return;
        dir = nextDir;
        var head = snake[0];
        var newHead = (
            (head.x + dir.x + SIZE) % SIZE,
            (head.y + dir.y + SIZE) % SIZE,
            (head.z + dir.z + SIZE) % SIZE
        );
        if (Contains(newHead))
        {
            gameOver = true;
            if (score > high) high = score;
            return;
        }
        snake.Insert(0, newHead);
        if (newHead == food)
        {
            score++;
            SpawnFood();
            speed = Math.Max(80, speed - 10);
        }
        else
        {
            snake.RemoveAt(snake.Count - 1);
        }
    }

    void Render()
    {
        Console.Clear();
        Console.WriteLine($"🐍 3D SNAKE   Score: {score}   High: {high}   Speed: {(speed/1000.0):F2}s");
        Console.WriteLine("Controls: W/A/S/D/Q/E | Space=pause | R=restart | Esc=quit");
        for (int y = SIZE - 1; y >= 0; y--)
        {
            for (int z = 0; z < SIZE; z++)
            {
                for (int x = 0; x < SIZE; x++)
                {
                    var pos = (x, y, z);
                    char ch = '.';
                    if (snake.Contains(pos)) ch = 'o';
                    if (snake[0] == pos) ch = 'O';
                    if (food == pos) ch = '★';
                    Console.Write($"{ch} ");
                }
                Console.WriteLine();
            }
        }
        if (paused) Console.WriteLine("\n⏸️  PAUSED");
        if (gameOver) Console.WriteLine("\n💀 GAME OVER! Press R to restart.");
    }

    void InputLoop()
    {
        while (running)
        {
            var key = Console.ReadKey(true);
            switch (key.Key)
            {
                case ConsoleKey.Spacebar:
                    paused = !paused;
                    break;
                case ConsoleKey.R:
                    if (gameOver)
                    {
                        int c = SIZE / 2;
                        snake.Clear();
                        snake.Add((c, c, c));
                        dir = (0, 0, 1);
                        nextDir = (0, 0, 1);
                        score = 0;
                        speed = 300;
                        gameOver = false;
                        SpawnFood();
                    }
                    break;
                case ConsoleKey.W: nextDir = (0, 0, 1); break;
                case ConsoleKey.S: nextDir = (0, 0, -1); break;
                case ConsoleKey.A: nextDir = (-1, 0, 0); break;
                case ConsoleKey.D: nextDir = (1, 0, 0); break;
                case ConsoleKey.Q: nextDir = (0, 1, 0); break;
                case ConsoleKey.E: nextDir = (0, -1, 0); break;
                case ConsoleKey.UpArrow: nextDir = (0, 0, 1); break;
                case ConsoleKey.DownArrow: nextDir = (0, 0, -1); break;
                case ConsoleKey.LeftArrow: nextDir = (-1, 0, 0); break;
                case ConsoleKey.RightArrow: nextDir = (1, 0, 0); break;
                case ConsoleKey.Escape:
                    running = false;
                    return;
            }
        }
    }

    void Run()
    {
        while (running && !gameOver)
        {
            Update();
            Render();
            Thread.Sleep(speed);
        }
        if (gameOver)
        {
            Render();
            while (running)
            {
                var key = Console.ReadKey(true);
                if (key.Key == ConsoleKey.R)
                {
                    int c = SIZE / 2;
                    snake.Clear();
                    snake.Add((c, c, c));
                    dir = (0, 0, 1);
                    nextDir = (0, 0, 1);
                    score = 0;
                    speed = 300;
                    gameOver = false;
                    SpawnFood();
                    Run();
                    return;
                }
                if (key.Key == ConsoleKey.Escape) break;
            }
        }
    }

    static void Main()
    {
        new Snake3D();
    }
}
